# Explorary Data Analysis 
В данном блокноте будет исследоваться датасет для обучения - [Bank Credit Risk Assessment Dataset](https://www.kaggle.com/datasets/kapturovalexander/bank-credit-risk-assessment).
В рамках данного проекта он находится в файле [data/test_data.csv](data/test_data.csv)

Explorary Data Analysis преследует следущие цели:
1) Изучение структуры и характеристик набора данных
2) Идентификация связий и корреляций между переменными
3) Подготовка данных для дальнейших этапов анализа

С этой целью, используя `pandas` и `matplotlib` мы будем исследовать предоставленный датасет и готовить его к дальнейшему использованию. Так же для визуализации будет использоваться библиотека `seaborn`.

# Шаг 0. Подготовка окружения.

In [1]:
import pandas as pd
import seaborn as sns

training_data = pd.read_csv('../data/train_data.csv')

# Шаг 1. Исследования табличного вида и состава данных.
Сперва нам необходимо понять что из себя представляет предоставленный датасет. Для этого посмотрим на его форму и содержание первых нескольких строк

In [2]:
training_data.shape # получаем размер выборки и кол-во параметров

(18420, 18)

In [3]:
training_data.head()

,Задолженность,"Просрочка, дни",Первоначльный лимит,BIRTHDATE,SEX,EDU,INCOME,TERM,Рейтинг кредитной истории,LV_AREA,LV_SETTLEMENTNAME,INDUSTRYNAME,PDN,CLIENTID,SCORINGMARK,UNDERAGECHILDRENCOUNT,VELCOMSCORING,FAMILYSTATUS
0,6063.50,3,7000.0,1983-07-08,Мужской,Среднее специальное,703.07,60,A1,Гомельская область,КОСАКОВКА,АПК,0.98,919517,233.0,0,NaN,1
1,3765.04,0,5000.0,1987-12-19,Женский,Высшее,1693.68,60,B2,NaN,МИНСК,Финансы и страхование,0.17,539353,NaN,1,NaN,1
2,2067.66,0,2650.0,1966-02-28,Женский,Среднее,724.49,60,C2,ГРОДНЕНСКАЯ,ЁДКИ,Медицина и здравоохранение,0.40,818913,NaN,0,NaN,2
3,2370.39,0,3000.0,1972-02-18,Женский,Среднее специальное,1045.84,60,A3,МИНСКАЯ,ФАНИПОЛЬ,Государственное управление,0.29,691598,216.0,0,NaN,2
4,2280.55,0,3000.0,1997-02-23,Женский,Высшее,1092.65,60,B3,NaN,Минск,Медицина и здравоохранение,0.63,425193,NaN,0,NaN,2


In [4]:
training_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18420 entries, 0 to 18419
Data columns (total 18 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Задолженность              18420 non-null  float64
 1   Просрочка, дни             18420 non-null  int64  
 2   Первоначльный лимит        18420 non-null  float64
 3   BIRTHDATE                  18420 non-null  object 
 4   SEX                        18420 non-null  object 
 5   EDU                        18420 non-null  object 
 6   INCOME                     18420 non-null  float64
 7   TERM                       18420 non-null  int64  
 8   Рейтинг кредитной истории  17488 non-null  object 
 9   LV_AREA                    16297 non-null  object 
 10  LV_SETTLEMENTNAME          18374 non-null  object 
 11  INDUSTRYNAME               18420 non-null  object 
 12  PDN                        18420 non-null  float64
 13  CLIENTID                   18420 non-null  int

In [5]:
training_data.describe()

,Задолженность,"Просрочка, дни",Первоначльный лимит,INCOME,TERM,PDN,CLIENTID,SCORINGMARK,UNDERAGECHILDRENCOUNT,VELCOMSCORING,FAMILYSTATUS
count,18420.000000,18420.000000,18420.000000,18420.000000,18420.000000,18420.000000,1.842000e+04,16786.000000,18420.000000,8889.000000,18420.000000
mean,4318.769750,9.196743,5127.834256,1664.840782,52.301737,0.290288,8.226245e+05,186.432622,0.636754,177.040162,1.498751
std,4336.432353,42.309092,4879.201691,1048.813447,13.378788,0.144359,1.966853e+05,67.430848,0.854167,310.624940,0.500988
min,0.000000,0.000000,180.000000,0.000000,23.000000,0.000000,1.289800e+04,-108.000000,0.000000,-2.000000,1.000000
25%,1287.900000,0.000000,1550.000000,1021.692500,36.000000,0.190000,7.193478e+05,134.000000,0.000000,-2.000000,1.000000
50%,2675.670000,0.000000,3200.000000,1460.175000,60.000000,0.300000,9.083645e+05,182.000000,0.000000,-2.000000,1.000000
75%,5785.870000,0.000000,6912.500000,1980.460000,60.000000,0.390000,9.621360e+05,234.000000,1.000000,278.000000,2.000000
max,40000.000000,491.000000,40000.000000,29942.820000,60.000000,1.210000,1.061453e+06,416.000000,3.000000,998.000000,4.000000


Из беглого анализа просмотра данных можно сделать следущие выводы:
- **Форма**
  - Датасет содержит 18 параметров, распределенных для 18420 записей.
  - Для параметров `Рейтинг кредитной истории`, `LV_AREA`, `LV_SETTLEMENTNAME`, `SCORINGMARK`, `VELCOMSCORING` некоторые записи пропущены.
  - Все числовые параметры (`Задолженность`, `Просрочка, дни`, `INCOME`, `TERM`, `PDN`, `UNDERAGECHILDRENCOUNT`, `FAMILYSTATUS`), которые могут повлиять на результат (кроме `SCORINGMARK` и `VELCOMSCORING`) заполнены значениями на 100%
  - `Рейтинг кредитной истории` имеет высокую, но не полную заполняемость.
- **Цели**
  - Поскольку цель этого проекта - скоринг (вероятность того, что заемщик отдаст кредит), а в датасете нет соответствующего поля, необходимо выбрать другое.
  - Целивыми полями назовем поля "Задолженность" (чем меньше, тем лучше). Будем считать, что пользователь отдал кредит в том случае, если его задолжность равна 0.
  - Необходимо выявить корреляцию между значениями. Сделаем это позже.
- **Мета**
  - Данные представлены для Беларуси. Поэтому кредитный рейтинг меряется в классах.
  - Так же в датасете есть специфичные для Беларуси параметры, такие как `VELCOMSCORING`.
  - На первый взгляд аномалий не наблюдается.

# Шаг 1. Обработка пропущенных значений

Из исследования формы можно сделать вывод, что датасет не полон. 
Для нивелирования влияния пропущенных данных необходимо исследовать пропущенные значения и решить, что необходимо с ними сделать.

Подходящие варианты:
1. Дроп пропущенных значений. Подходит, если их не много (не превышает стандартной погрешности в 10%).
2. Заполнение. Пропущенные данные можно заполнить медианным или средним значениями. **Это может повлиять на выборку.**
3. Дроп параметра. Если пропущенные данные сконцентрированы в одном параметре, можно сделать предположение, что он не значителен, и пренибречь им.

In [6]:
training_data.isnull().sum()

Задолженность                   0
Просрочка, дни                  0
Первоначльный лимит             0
BIRTHDATE                       0
SEX                             0
EDU                             0
INCOME                          0
TERM                            0
Рейтинг кредитной истории     932
LV_AREA                      2123
LV_SETTLEMENTNAME              46
INDUSTRYNAME                    0
PDN                             0
CLIENTID                        0
SCORINGMARK                  1634
UNDERAGECHILDRENCOUNT           0
VELCOMSCORING                9531
FAMILYSTATUS                    0
dtype: int64

Видно, что основная часть пропусков сконцентрирована в 3 столбцах: `LV_AREA`, `SCORINGMARK`, `VELCOMSCORING`.
- Для `LV_AREA` можно сделать предположение о несущественности и не принимать во внимание этот параметр.**Дропаем.**
- Для `SCORINGMARK` очевидно, что не все пользователи имеют кредитную историю. Однако, этот параметр нельзя не принимать во внимание.
- `VELCOMSCORING` - специфичный для Беларуси параметр. Его можно **смело дропать**.

Количество пропусков в полях `Рейтинг кредитной истории` и `LV_SETTLEMENTNAME` является незначительным относительно выборки. **Дропаем** строчки с пропущенными значениями.
Что касается `SCORINGMARK`, заполним его минимальным значением. Это решение **принято волюнтарно** из тех соображений, что в реальности людям без кредитной истории менее вероятно дают кредит.

Все подчистки данных выполняются в файле [missing_values.py](./data_preparation/missing_values.py). Импортируем необходимые функции и исполним их:

In [7]:
from data_preparation.missing_values import clear_missing_values, paste_min_value_to_columns

training_data = clear_missing_values(training_data, columns=['LV_AREA', 'VELCOMSCORING'], empty_rows_on=['Рейтинг кредитной истории', 'LV_SETTLEMENTNAME'])
training_data = paste_min_value_to_columns(training_data, ['SCORINGMARK'])



Проверим результат обработки пропущенных значений:

In [8]:
training_data.isnull().sum()

Задолженность                0
Просрочка, дни               0
Первоначльный лимит          0
BIRTHDATE                    0
SEX                          0
EDU                          0
INCOME                       0
TERM                         0
Рейтинг кредитной истории    0
LV_SETTLEMENTNAME            0
INDUSTRYNAME                 0
PDN                          0
CLIENTID                     0
SCORINGMARK                  0
UNDERAGECHILDRENCOUNT        0
FAMILYSTATUS                 0
dtype: int64